# Introduction

In this notebook, we will do a comprehensive analysis of the Android app market by comparing thousands of apps in the Google Play store.

# About the Dataset of Google Play Store Apps & Reviews

**Data Source:** <br>
App and review data was scraped from the Google Play Store by Lavanya Gupta in 2018. Original files listed [here](
https://www.kaggle.com/lava18/google-play-store-apps).

# Import Statements

In [256]:
import pandas as pd


# Notebook Presentation

In [257]:
# Show numeric output in decimal format e.g., 2.15
pd.options.display.float_format = '{:,.2f}'.format

# Read the Dataset

In [258]:
df_apps = pd.read_csv('apps.csv')

# Data Cleaning

**Challenge**: How many rows and columns does `df_apps` have? What are the column names? Look at a random sample of 5 different rows with [.sample()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html).

In [259]:
df_apps.shape

(10841, 12)

In [260]:
df_apps.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size_MBs', 'Installs', 'Type',
       'Price', 'Content_Rating', 'Genres', 'Last_Updated', 'Android_Ver'],
      dtype='object')

In [261]:
df_apps.head()

App         Category  Rating  \
0                       Ak Parti Yardım Toplama           SOCIAL     NaN   
1                    Ain Arabic Kids Alif Ba ta           FAMILY     NaN   
2  Popsicle Launcher for Android P 9.0 launcher  PERSONALIZATION     NaN   
3                     Command & Conquer: Rivals           FAMILY     NaN   
4                                    CX Network         BUSINESS     NaN   

   Reviews  Size_MBs Installs  Type   Price Content_Rating           Genres  \
0        0      8.70        0  Paid  $13.99           Teen           Social   
1        0     33.00        0  Paid   $2.99       Everyone        Education   
2        0      5.50        0  Paid   $1.49       Everyone  Personalization   
3        0     19.00        0   NaN       0   Everyone 10+         Strategy   
4        0     10.00        0  Free       0       Everyone         Business   

     Last_Updated         Android_Ver  
0   July 28, 2017          4.1 and up  
1  April 15, 2016          3.0 and up  
2   July 11, 2018          4.2 and up  
3   June 28, 2018  Varies with device  
4  August 6, 2018          4.1 and up

In [262]:
df_apps.sample(5)

App             Category  \
6824  CarMax – Cars for Sale: Search Used Car Inventory    AUTO_AND_VEHICLES   
2244                                            DM 24/7                TOOLS   
9757                                    Baby Panda Care               FAMILY   
961                                           AW CAMERA                TOOLS   
8399                    Citymapper - Transit Navigation  MAPS_AND_NAVIGATION   

      Rating  Reviews  Size_MBs    Installs  Type Price Content_Rating  \
6824    4.40    21777      9.00   1,000,000  Free     0       Everyone   
2244    3.60        7     16.00       1,000  Free     0       Everyone   
9757    4.20   108795     49.00  10,000,000  Free     0       Everyone   
961      NaN        2     24.00         100  Free     0       Everyone   
8399    4.50    65448      6.50   5,000,000  Free     0       Everyone   

                        Genres      Last_Updated         Android_Ver  
6824           Auto & Vehicles    August 4, 2018  Varies with device  
2244                     Tools    August 1, 2018        2.3.3 and up  
9757  Educational;Pretend Play     June 15, 2018        4.0.3 and up  
961                      Tools  January 30, 2018          2.3 and up  
8399         Maps & Navigation    August 2, 2018  Varies with device

### Drop Unused Columns

**Challenge**: Remove the columns called `Last_Updated` and `Android_Ver` from the DataFrame. We will not use these columns.

In [263]:
df_apps.drop(columns=['Last_Updated', 'Android_Ver'], axis=1, inplace=True)
df_apps.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size_MBs', 'Installs', 'Type',
       'Price', 'Content_Rating', 'Genres'],
      dtype='object')

### Find and Remove NaN values in Ratings

**Challenge**: How may rows have a NaN value (not-a-number) in the Ratings column? Create DataFrame called `df_apps_clean` that does not include these rows.

In [264]:
df_apps.isna().any()

App               False
Category          False
Rating             True
Reviews           False
Size_MBs          False
Installs          False
Type               True
Price             False
Content_Rating    False
Genres            False
dtype: bool

In [265]:
nan_rows = df_apps[df_apps['Rating'].isna()]
nan_rows.shape

(1474, 10)

In [266]:
df_apps_clean = df_apps.dropna()
df_apps_clean.isna().any()

App               False
Category          False
Rating            False
Reviews           False
Size_MBs          False
Installs          False
Type              False
Price             False
Content_Rating    False
Genres            False
dtype: bool

In [267]:
df_apps_clean.shape

(9367, 10)

### Find and Remove Duplicates

**Challenge**: Are there any duplicates in data? Check for duplicates using the [.duplicated()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html) function. How many entries can you find for the "Instagram" app? Use [.drop_duplicates()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html) to remove any duplicates from `df_apps_clean`.


In [268]:
df_apps_clean.duplicated().value_counts()

False    8891
True      476
dtype: int64

In [269]:
duplicated_rows = df_apps_clean[df_apps_clean.duplicated()]
duplicated_rows.shape

(476, 10)

In [270]:
df_apps_clean[df_apps_clean['App'] == 'Instagram']

App Category  Rating   Reviews  Size_MBs       Installs  Type  \
10806  Instagram   SOCIAL    4.50  66577313      5.30  1,000,000,000  Free   
10808  Instagram   SOCIAL    4.50  66577446      5.30  1,000,000,000  Free   
10809  Instagram   SOCIAL    4.50  66577313      5.30  1,000,000,000  Free   
10810  Instagram   SOCIAL    4.50  66509917      5.30  1,000,000,000  Free   

      Price Content_Rating  Genres  
10806     0           Teen  Social  
10808     0           Teen  Social  
10809     0           Teen  Social  
10810     0           Teen  Social

We can see that there are 4 entries for Instagram and the only column with a difference is the Reviews column. If we simply call the .drop_duplicates() 3 copies will be kept anyways and that's not what we want. We need to specify the names of the columns (a subset) that should be used to determine duplicates.

In [271]:
df_apps_clean = df_apps_clean.drop_duplicates(subset=['App', 'Type', 'Price'])
df_apps_clean.duplicated().any()

False

In [272]:
df_apps_clean.shape

(8199, 10)

# Find Highest Rated Apps

**Challenge**: Identify which apps are the highest rated. What problem might you encounter if you rely exclusively on ratings alone to determine the quality of an app?

In [273]:
highest_rated_apps = df_apps_clean[df_apps_clean['Rating'] == df_apps_clean['Rating'].max()]
highest_rated_apps.head()

App Category  Rating  Reviews  \
21                              KBA-EZ Health Guide  MEDICAL    5.00        4   
28                                         Ra Ga Ba     GAME    5.00        2   
47                                          Mu.F.O.     GAME    5.00        2   
82                                 Brick Breaker BR     GAME    5.00        7   
99  Anatomy & Physiology Vocabulary Exam Review App  MEDICAL    5.00        1   

    Size_MBs Installs  Type  Price Content_Rating   Genres  
21     25.00        1  Free      0       Everyone  Medical  
28     20.00        1  Paid  $1.49       Everyone   Arcade  
47     16.00        1  Paid  $0.99       Everyone   Arcade  
82     19.00        5  Free      0       Everyone   Arcade  
99      4.60        5  Free      0       Everyone  Medical

In [274]:
highest_rated_apps.shape

(271, 10)

# Find 5 Largest Apps in terms of Size (MBs)

**Challenge**: What's the size in megabytes (MB) of the largest Android apps in the Google Play Store. Based on the data, do you think there could be limit in place or can developers make apps as large as they please?

In [275]:
largest_apps = df_apps_clean[df_apps_clean['Size_MBs'] == df_apps_clean['Size_MBs'].max()]
largest_apps.head()

App            Category  \
1795                              Navi Radiography Pro             MEDICAL   
3144                                        Vi Trainer  HEALTH_AND_FITNESS   
4176  Car Crash III Beam DH Real Damage Simulator 2018                GAME   
7926                                         Post Bank             FINANCE   
7927                       The Walking Dead: Our World                GAME   

      Rating  Reviews  Size_MBs   Installs  Type   Price Content_Rating  \
1795    4.70       11    100.00        500  Paid  $15.99       Everyone   
3144    3.60      124    100.00      5,000  Free       0       Everyone   
4176    3.60      151    100.00     10,000  Free       0       Everyone   
7926    4.50    60449    100.00  1,000,000  Free       0       Everyone   
7927    4.00    22435    100.00  1,000,000  Free       0           Teen   

                Genres  
1795           Medical  
3144  Health & Fitness  
4176            Racing  
7926           Finance  
7927            Action

In [276]:
largest_apps.shape

(14, 10)

In [277]:
print(f'The largest app size is {largest_apps["Size_MBs"].max()} MB.')

The largest app size is 100.0 MB.


# Find the 5 App with Most Reviews

**Challenge**: Which apps have the highest number of reviews? Are there any paid apps among the top 50?

In [278]:
df_apps_clean[df_apps_clean['Reviews'] == df_apps_clean['Reviews'].max()]

App Category  Rating   Reviews  Size_MBs       Installs  Type  \
10805  Facebook   SOCIAL    4.10  78158306      5.30  1,000,000,000  Free   

      Price Content_Rating  Genres  
10805     0           Teen  Social

In [279]:
most_reviewd_apps = df_apps_clean.sort_values(by=['Reviews'], ascending=False).head(50)
most_reviewd_apps

App            Category  \
10805                                           Facebook              SOCIAL   
10785                                 WhatsApp Messenger       COMMUNICATION   
10806                                          Instagram              SOCIAL   
10784           Messenger – Text and Video Chat for Free       COMMUNICATION   
10650                                     Clash of Clans                GAME   
10744            Clean Master- Space Cleaner & Antivirus               TOOLS   
10835                                     Subway Surfers                GAME   
10828                                            YouTube       VIDEO_PLAYERS   
10746  Security Master - Antivirus, VPN, AppLock, Boo...               TOOLS   
10584                                       Clash Royale                GAME   
10763                                   Candy Crush Saga                GAME   
10770        UC Browser - Fast Download Private & Secure       COMMUNICATION   
10735                                           Snapchat              SOCIAL   
10489    360 Security - Free Antivirus, Booster, Cleaner               TOOLS   
10731                                     My Talking Tom                GAME   
10594                                        8 Ball Pool                GAME   
10302  DU Battery Saver - Battery Charger & Battery Life               TOOLS   
10354                        BBM - Free Calls & Messages       COMMUNICATION   
10549  Cache Cleaner-DU Speed Booster (booster & clea...               TOOLS   
10757                                            Twitter  NEWS_AND_MAGAZINES   
10721                                    Viber Messenger       COMMUNICATION   
10578                                     Shadow Fight 2                GAME   
10813                                      Google Photos         PHOTOGRAPHY   
10724                        LINE: Free Calls & Messages       COMMUNICATION   
10717                                                Pou                GAME   
10792                      Skype - free IM & video calls       COMMUNICATION   
10628                                         Pokémon GO                GAME   
10388           Minion Rush: Despicable Me Official Game                GAME   
10694                                            Yes day                GAME   
10695                                            Hay Day              FAMILY   
10644                           Dream League Soccer 2018                GAME   
10696                                  My Talking Angela                GAME   
10660             VivaVideo - Video Editor & Photo Movie       VIDEO_PLAYERS   
10786                       Google Chrome: Fast & Secure       COMMUNICATION   
10817                          Maps - Navigate & Explore    TRAVEL_AND_LOCAL   
10672                                  Hill Climb Racing                GAME   
10734                                      Facebook Lite              SOCIAL   
10649                                Asphalt 8: Airborne                GAME   
10699                          Mobile Legends: Bang Bang                GAME   
10322  Battery Doctor-Battery Life Saver & Battery Co...               TOOLS   
10396                                     Piano Tiles 2™                GAME   
10777                                       Temple Run 2                GAME   
10822                                             Google               TOOLS   
10359  Truecaller: Caller ID, SMS spam blocking & Dialer       COMMUNICATION   
10711                         SHAREit - Transfer & Share               TOOLS   
10389   Sniper 3D Gun Shooter: Free Shooting Games - FPS                GAME   
10676                                   Farm Heroes Saga                GAME   
10576   PicsArt Photo Studio: Collage Maker & Pic Editor         PHOTOGRAPHY   
10461  PhotoGrid: Video & Pic Collage Maker, Photo Ed...         PHOTOGRAPHY   
10502   GO Launcher - 3D parallax Themes & HD Wallpap

# Plotly Pie and Donut Charts - Visualise Categorical Data: Content Ratings

All Android apps have a content rating like “Everyone” or “Teen” or “Mature 17+”. Let’s take a look at the distribution of the content ratings in our dataset and see how to visualise it with plotly - a popular data visualisation library that you can use alongside or instead of Matplotlib.

First, we’ll count the number of occurrences of each rating with .value_counts()

In [280]:
ratings = df_apps_clean['Content_Rating'].value_counts()
ratings

Everyone           6621
Teen                912
Mature 17+          357
Everyone 10+        305
Adults only 18+       3
Unrated               1
Name: Content_Rating, dtype: int64

The first step in creating charts with plotly is to import plotly.express. This is the fastest way to create a beautiful graphic with a minimal amount of code in plotly.

In [281]:
import plotly.express as px

To create a pie chart we simply call px.pie() and then .show() the resulting figure. Plotly refers to all their figures, be they line charts, bar charts, or pie charts as graph_objects.

In [282]:
fig = px.pie(labels=ratings.index, values=ratings.values)
fig.show()

/usr/local/lib/python3.10/dist-packages/plotly/express/_core.py:137: FutureWarning:

Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.



Let’s customise our pie chart. Looking at the .pie() [documentation](https://https://plotly.com/python-api-reference/generated/plotly.express.pie.html) we see a number of parameters that we can set, like title or names.

If you’d like to configure other aspects of the chart, that you can’t see in the list of parameters, you can call a method called .update_traces(). In plotly lingo, “traces” refer to graphical marks on a figure. Think of “traces” as collections of attributes. Here we update the traces to change how the text is displayed.

In [283]:
fig = px.pie(labels=ratings.index, values=ratings.values, title='Content Rating', names=ratings.index)
fig.update_traces(textposition='outside', textinfo='percent+label')
fig.show()

/usr/local/lib/python3.10/dist-packages/plotly/express/_core.py:137: FutureWarning:

Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.



To create a donut 🍩 chart, we can simply add a value for the hole argument:

In [284]:
fig = px.pie(labels=ratings.index, values=ratings.values, title='Content Rating', names=ratings.index, hole=0.6)
fig.update_traces(textposition='outside', textinfo='percent+label', textfont_size=15)
fig.show()

/usr/local/lib/python3.10/dist-packages/plotly/express/_core.py:137: FutureWarning:

Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.



There is a different way you can create the pie chart, without having to create a new Series like we did when using .value_counts(). You can directly put the df as an argument with the column name as value for the 'names' key and it will automatically group the rows with the same value, like so.

In [285]:
fig = px.pie(data_frame=df_apps_clean, names='Content_Rating', title='Content Rating', hole=0.6)
fig.update_traces(textposition='outside', textinfo='percent+label', textfont_size=15)
fig.show()

# Numeric Type Conversion: Examine the Number of Installs

**Challenge**: How many apps had over 1 billion (that's right - BILLION) installations? How many apps just had a single install?

Check the datatype of the Installs column.

Count the number of apps at each level of installations.

Convert the number of installations (the Installs column) to a numeric data type. Hint: this is a 2-step process. You'll have make sure you remove non-numeric characters first.

In [286]:
print(type(df_apps_clean['Installs'].max()))

<class 'str'>


In [287]:
df_apps_clean.value_counts(subset=['Installs'])

Installs     
1,000,000        1417
100,000          1096
10,000            988
10,000,000        933
1,000             698
5,000,000         607
500,000           504
50,000            457
5,000             425
100               303
50,000,000        202
500               199
100,000,000       189
10                 69
50                 56
500,000,000        24
1,000,000,000      20
5                   9
1                   3
dtype: int64

In [288]:
df_apps_clean['Installs'] = df_apps_clean['Installs'].astype(str).str.replace(',', '')
df_apps_clean['Installs'] = pd.to_numeric(df_apps_clean['Installs'])
print(type(df_apps_clean['Installs'].max()))

<class 'numpy.int64'>


In [289]:
df_apps_clean.value_counts(subset=['Installs']).sort_index(ascending=False)

Installs  
1000000000      20
500000000       24
100000000      189
50000000       202
10000000       933
5000000        607
1000000       1417
500000         504
100000        1096
50000          457
10000          988
5000           425
1000           698
500            199
100            303
50              56
10              69
5                9
1                3
dtype: int64

In [290]:
# Alternative
df_apps_clean[['App', 'Installs']].groupby('Installs').count().sort_index(ascending=False)

App
Installs        
1000000000    20
500000000     24
100000000    189
50000000     202
10000000     933
5000000      607
1000000     1417
500000       504
100000      1096
50000        457
10000        988
5000         425
1000         698
500          199
100          303
50            56
10            69
5              9
1              3

# Find the Most Expensive Apps, Filter out the Junk, and Calculate a (ballpark) Sales Revenue Estimate

Let's examine the Price column more closely.

**Challenge**: Convert the price column to numeric data. Then investigate the top 20 most expensive apps in the dataset.

Remove all apps that cost more than $250 from the `df_apps_clean` DataFrame.

Add a column called 'Revenue_Estimate' to the DataFrame. This column should hold the price of the app times the number of installs. What are the top 10 highest grossing paid apps according to this estimate? Out of the top 10 highest grossing paid apps, how many are games?


In [291]:
print(type(df_apps_clean['Price'].max()))

<class 'str'>


In [292]:
df_apps_clean['Price'] = df_apps_clean['Price'].astype(str).str.replace('$', '')
df_apps_clean['Price'] = pd.to_numeric(df_apps_clean['Price'])
print(type(df_apps_clean['Price'].max()))

<class 'numpy.float64'>


<ipython-input-292-9832efbf8d16>:1: FutureWarning:

The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.



In [293]:
df_apps_clean.sort_values('Price', ascending=False).head(20)

App   Category  Rating  Reviews  Size_MBs  \
3946        I'm Rich - Trump Edition  LIFESTYLE    3.60      275      7.30   
2461              I AM RICH PRO PLUS    FINANCE    4.00       36     41.00   
4606               I Am Rich Premium    FINANCE    4.10     1867      4.70   
3145              I am rich(premium)    FINANCE    3.50      472      0.94   
3554                      💎 I'm rich  LIFESTYLE    3.80      718     26.00   
5765                       I am rich  LIFESTYLE    3.80     3547      1.80   
1946  I am rich (Most expensive app)    FINANCE    4.10      129      2.70   
2775                   I Am Rich Pro     FAMILY    4.40      201      2.70   
3221                  I am Rich Plus     FAMILY    4.00      856      8.70   
3114                       I am Rich    FINANCE    4.30      180      3.80   
1331          most expensive app (H)     FAMILY    4.30        6      1.50   
2394                      I am Rich!    FINANCE    3.80       93     22.00   
3897                       I Am Rich     FAMILY    3.60      217      4.90   
2193             I am extremely Rich  LIFESTYLE    2.90       41      2.90   
3856                   I am rich VIP  LIFESTYLE    3.80      411      2.60   
2281       Vargo Anesthesia Mega App    MEDICAL    4.60       92     32.00   
1407                    LTC AS Legal    MEDICAL    4.00        6      1.30   
2629                I am Rich Person  LIFESTYLE    4.20      134      1.80   
2481         A Manual of Acupuncture    MEDICAL    3.50      214     68.00   
4264         Golfshot Plus: Golf GPS     SPORTS    4.10     3387     25.00   

      Installs  Type  Price Content_Rating         Genres  
3946     10000  Paid 400.00       Everyone      Lifestyle  
2461      1000  Paid 399.99       Everyone        Finance  
4606     50000  Paid 399.99       Everyone        Finance  
3145      5000  Paid 399.99       Everyone        Finance  
3554     10000  Paid 399.99       Everyone      Lifestyle  
5765    100000  Paid 399.99       Everyone      Lifestyle  
1946      1000  Paid 399.99           Teen        Finance  
2775      5000  Paid 399.99       Everyone  Entertainment  
3221     10000  Paid 399.99       Everyone  Entertainment  
3114      5000  Paid 399.99       Everyone        Finance  
1331       100  Paid 399.99       Everyone  Entertainment  
2394      1000  Paid 399.99       Everyone        Finance  
3897     10000  Paid 389.99       Everyone  Entertainment  
2193      1000  Paid 379.99       Everyone      Lifestyle  
3856     10000  Paid 299.99       Everyone      Lifestyle  
2281      1000  Paid  79.99       Everyone        Medical  
1407       100  Paid  39.99       Everyone        Medical  
2629      1000  Paid  37.99       Everyone      Lifestyle  
2481      1000  Paid  33.99       Everyone        Medical  
4264     50000  Paid  29.99       Everyone         Sports

In [294]:
df_apps_clean[['Price', 'App']].groupby('Price').count().sort_index(ascending=False)

App
Price       
400.00     1
399.99    11
389.99     1
379.99     1
299.99     1
...      ...
1.29       1
1.20       1
1.00       2
0.99     104
0.00    7595

[73 rows x 1 columns]

Drop apps over $250.

In [295]:
# ALTERNATIVE, must add .index after selecting the rows
# df_apps_clean.drop(df_apps_clean[df_apps_clean['Price'] > 250.00].index, inplace=True)
# df_apps_clean.sort_values('Price', ascending=False)

In [296]:
# ALTERNATIVE 2
# df_apps_clean = df_apps_clean.query('Price <= 250.00')
# df_apps_clean.sort_values('Price', ascending=False)

In [297]:
df_apps_clean = df_apps_clean[df_apps_clean['Price'] <= 250.00]
df_apps_clean.sort_values('Price', ascending=False)

App   Category  Rating   Reviews  Size_MBs  \
2281     Vargo Anesthesia Mega App    MEDICAL    4.60        92     32.00   
1407                  LTC AS Legal    MEDICAL    4.00         6      1.30   
2629              I am Rich Person  LIFESTYLE    4.20       134      1.80   
2481       A Manual of Acupuncture    MEDICAL    3.50       214     68.00   
2463            PTA Content Master    MEDICAL    4.20        64     41.00   
...                            ...        ...     ...       ...       ...   
4508   myAir™ for Air10™ by ResMed    MEDICAL    3.70       236     18.00   
4507                 AK Math Coach     FAMILY    3.60       283     18.00   
4506          Forgotten Hill: Fall       GAME    4.40      1063     18.00   
4505                AE Video Poker       GAME    4.00       721     18.00   
10835               Subway Surfers       GAME    4.50  27722264     76.00   

         Installs  Type  Price Content_Rating     Genres  
2281         1000  Paid  79.99       Everyone    Medical  
1407          100  Paid  39.99       Everyone    Medical  
2629         1000  Paid  37.99       Everyone  Lifestyle  
2481         1000  Paid  33.99       Everyone    Medical  
2463         1000  Paid  29.99       Everyone    Medical  
...           ...   ...    ...            ...        ...  
4508        50000  Free   0.00       Everyone    Medical  
4507        50000  Free   0.00       Everyone  Education  
4506        50000  Free   0.00           Teen  Adventure  
4505        50000  Free   0.00           Teen     Casino  
10835  1000000000  Free   0.00   Everyone 10+     Arcade  

[8184 rows x 10 columns]

### The most expensive apps sub $250

In [298]:
df_apps_clean.sort_values('Price', ascending=False).head()

App   Category  Rating  Reviews  Size_MBs  \
2281  Vargo Anesthesia Mega App    MEDICAL    4.60       92     32.00   
1407               LTC AS Legal    MEDICAL    4.00        6      1.30   
2629           I am Rich Person  LIFESTYLE    4.20      134      1.80   
2481    A Manual of Acupuncture    MEDICAL    3.50      214     68.00   
2463         PTA Content Master    MEDICAL    4.20       64     41.00   

      Installs  Type  Price Content_Rating     Genres  
2281      1000  Paid  79.99       Everyone    Medical  
1407       100  Paid  39.99       Everyone    Medical  
2629      1000  Paid  37.99       Everyone  Lifestyle  
2481      1000  Paid  33.99       Everyone    Medical  
2463      1000  Paid  29.99       Everyone    Medical

### Highest Grossing Paid Apps (ballpark estimate)

We can work out the highest grossing paid apps now. All we need to do is multiply the values in the price and the installs column to get the number:

In [299]:
df_apps_clean['Revenue_Estimate'] = df_apps_clean['Price'] * df_apps_clean['Installs']
df_apps_clean.sort_values('Revenue_Estimate', ascending=False).head(10)

<ipython-input-299-b03f432743c4>:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



App     Category  Rating  Reviews  Size_MBs  \
9220                      Minecraft       FAMILY    4.50  2376564     19.00   
8825                  Hitman Sniper         GAME    4.60   408292     29.00   
7151  Grand Theft Auto: San Andreas         GAME    4.40   348962     26.00   
7477            Facetune - For Free  PHOTOGRAPHY    4.40    49553     48.00   
7977        Sleep as Android Unlock    LIFESTYLE    4.50    23966      0.85   
6594            DraStic DS Emulator         GAME    4.60    87766     12.00   
6082                   Weather Live      WEATHER    4.50    76593      4.75   
7954                    Bloons TD 5       FAMILY    4.60   190086     94.00   
7633        Five Nights at Freddy's         GAME    4.60   100805     50.00   
6746     Card Wars - Adventure Time       FAMILY    4.30   129603     23.00   

      Installs  Type  Price Content_Rating                     Genres  \
9220  10000000  Paid   6.99   Everyone 10+  Arcade;Action & Adventure   
8825  10000000  Paid   0.99     Mature 17+                     Action   
7151   1000000  Paid   6.99     Mature 17+                     Action   
7477   1000000  Paid   5.99       Everyone                Photography   
7977   1000000  Paid   5.99       Everyone                  Lifestyle   
6594   1000000  Paid   4.99       Everyone                     Action   
6082    500000  Paid   5.99       Everyone                    Weather   
7954   1000000  Paid   2.99       Everyone                   Strategy   
7633   1000000  Paid   2.99           Teen                     Action   
6746   1000000  Paid   2.99   Everyone 10+    Card;Action & Adventure   

      Revenue_Estimate  
9220     69,900,000.00  
8825      9,900,000.00  
7151      6,990,000.00  
7477      5,990,000.00  
7977      5,990,000.00  
6594      4,990,000.00  
6082      2,995,000.00  
7954      2,990,000.00  
7633      2,990,000.00  
6746      2,990,000.00

This generously assumes of course that all the installs would have been made at the listed price, which is unlikely, as there are always promotions and free give-aways on the App Stores.

# Plotly Bar Charts & Scatter Plots: Analysing App Categories

If you were to release an app, would you choose to go after a competitive category with many other apps? Or would you target a popular category with a high number of downloads? Or perhaps you can target a category which is both popular but also one where the downloads are spread out among many different apps. That way, even if it’s more difficult to discover among all the other apps, your app has a better chance of getting installed, right? Let’s analyse this with bar charts and scatter plots and figure out which categories are dominating the market.


We can find the number of different categories like so:

In [300]:
df_apps_clean['Category'].nunique()

33

To calculate the number of apps per category we can use our old friend .value_counts()

In [301]:
top_10_category = df_apps_clean['Category'].value_counts()[:10]
top_10_category

FAMILY             1606
GAME                910
TOOLS               719
PRODUCTIVITY        301
PERSONALIZATION     298
LIFESTYLE           297
FINANCE             296
MEDICAL             292
PHOTOGRAPHY         263
BUSINESS            262
Name: Category, dtype: int64

### Vertical Bar Chart - Highest Competition (Number of Apps)

To visualise this data in a bar chart we can use the plotly express (our px) bar() function:

In [302]:
bar = px.bar(x=top_10_category.index, y=top_10_category.values)
bar.show()

### Horizontal Bar Chart - Most Popular Categories (Highest Downloads)

But what if we look at it from a different perspective? What matters is not just the total number of apps in the category but how often apps are downloaded in that category. This will give us an idea of how popular a category is. First, we have to group all our apps by category and sum the number of installations.

In [303]:
category_installs = df_apps_clean.groupby('Category').sum('Installs')
category_installs.sort_values('Installs', ascending=True, inplace=True)
category_installs.head()

Rating  Reviews  Size_MBs  Installs  Price  Revenue_Estimate
Category                                                                
EVENTS      199.60   160590    592.53  15949410   0.00              0.00
BEAUTY      179.70   395133    619.50  26916200   0.00              0.00
PARENTING   215.00   953609  1,009.80  31116110   9.58        249,959.00
MEDICAL   1,217.70  1286805  5,302.49  39162676 623.87      5,872,355.00
COMICS      225.80  2340740    689.28  44931100   0.00              0.00

Then we can create a horizontal bar chart, simply by adding the orientation parameter (also setting the x axis as the 'Installs' columns and the y axis as the index).

In [304]:
h_bar = px.bar(x=category_installs['Installs'], y=category_installs.index, orientation='h')
h_bar.show()

We can also add a custom title and axis labels like using the .update_layouts() method.

In [305]:
h_bar = px.bar(x=category_installs['Installs'], y=category_installs.index, orientation='h', title='Category Popularity')
h_bar.update_layout(xaxis_title='Number of Downloads', yaxis_title='Category')
h_bar.show()

Now we see that Games and Tools are actually the most popular categories. If we plot the popularity of a category next to the number of apps in that category we can get an idea of how concentrated a category is. Do few apps have most of the downloads or are the downloads spread out over many apps?

### Category Concentration - Downloads vs. Competition

**Challenge**:
* First, create a DataFrame that has the number of apps in one column and the number of installs in another:

<img src=https://imgur.com/uQRSlXi.png width="350">

* Then use the [plotly express examples from the documentation](https://plotly.com/python/line-and-scatter/) alongside the [.scatter() API reference](https://plotly.com/python-api-reference/generated/plotly.express.scatter.html)to create scatter plot that looks like this.

<img src=https://imgur.com/cHsqh6a.png>

*Hint*: Use the size, hover_name and color parameters in .scatter(). To scale the yaxis, call .update_layout() and specify that the yaxis should be on a log-scale like so: yaxis=dict(type='log')

In [306]:
cat_merged_df = df_apps_clean.groupby('Category').count()[['App']]
cat_merged_df['Installs'] = category_installs['Installs']
cat_merged_df.sort_values('Installs', ascending=False, inplace=True)
cat_merged_df.head(6)

App     Installs
Category                       
GAME           910  13858762717
COMMUNICATION  257  11039241530
TOOLS          719   8099724500
PRODUCTIVITY   301   5788070180
SOCIAL         203   5487841475
PHOTOGRAPHY    263   4649143130

In [307]:
# ALTERNATIVE
# cat_number = df_apps_clean.groupby('Category').agg({'App': pd.Series.count})
# cat_merged_df = pd.merge(cat_number, category_installs['Installs'], on='Category', how='inner').sort_values('Installs', ascending=False)
# cat_merged_df.head(6)

In [308]:
fig = px.scatter(data_frame=cat_merged_df,
                 x='App', y='Installs',
                 title='Category Concentration',
                 size='App', color='Installs',
                 hover_name=cat_merged_df.index)
fig.update_layout(yaxis=dict(type='log'), xaxis_title='Number of Apps (Lower=More Concentrated)', yaxis_title='Installs')
fig.show()

What we see is that the categories like Family, Tools, and Game have many different apps sharing a high number of downloads. But for the categories like video players and entertainment, all the downloads are concentrated in very few apps.

<img src="https://img-c.udemycdn.com/redactor/raw/2020-10-11_13-16-11-a310e773b06e1efc0ad4114a12a51e01.png">

ALTERNATIVE SOLUTION:

For a plot that's supposed to show "concentration" (or installs/app) there's no single indicator that does that in Angela's version.  I changed my code to add an Installs/App column to the dataframe and have the color represent that variable.  This way you can easily see which categories are the most concentrated.

In [309]:
category_num = df_apps_clean.groupby("Category").agg({"App": pd.Series.count})
cat_df_merged2 = pd.merge(category_num, category_installs, on="Category", how="inner")
cat_df_merged2.sort_values("Installs", ascending=False, inplace=True)
cat_df_merged2['Installs/App'] = cat_df_merged2.Installs.div(cat_df_merged2.App)

px.scatter()
plot = px.scatter(cat_df_merged2, x="App",
                  y="Installs",
                  log_y=True,
                  color='Installs/App',
                  color_continuous_scale=px.colors.sequential.Plasma_r,
                  size=cat_df_merged2.App,
                  size_max=40,
                  hover_name=cat_df_merged2.index,
                  title='Category Concentration'
                  )
plot.update_layout(xaxis_title="Number of Apps",
                   yaxis_title="Installs",
                   )
plot.show()

# Extracting Nested Data from a Column

**Challenge**: How many different types of genres are there? Can an app belong to more than one genre? Check what happens when you use .value_counts() on a column with nested values? See if you can work around this problem by using the .split() function and the DataFrame's [.stack() method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.stack.html).


If we look at the number of unique values in the Genres column we get 114. But this is not accurate if we have nested data like we do here. We can see this using .value_counts() and looking at the values that just have a single entry. There we see that the semi-colon (;) separates the genre names.

In [310]:
len(df_apps_clean['Genres'].unique())

114

In [311]:
df_apps_clean['Genres'].value_counts().sort_values()[:5]

Lifestyle;Pretend Play      1
Strategy;Education          1
Adventure;Education         1
Role Playing;Brain Games    1
Tools;Education             1
Name: Genres, dtype: int64

<img src="https://img-c.udemycdn.com/redactor/raw/2020-10-11_13-18-53-e5af936a84354a54fc2ef2d1eea24e77.png">

We somehow need to separate the genre names to get a clear picture. This is where the string’s .split() method comes in handy.

In [312]:
stack = df_apps_clean.Genres.str.split(';')
stack.sample(30, random_state=30)

7690           [Casual, Education]
5735                       [Tools]
1409                      [Sports]
10306    [Video Players & Editors]
2295                    [Shopping]
5419            [Health & Fitness]
2666                      [Action]
6993                      [Social]
2511                   [Education]
10347              [Communication]
2159                [Art & Design]
7184                       [Tools]
7548               [Entertainment]
5131           [Maps & Navigation]
8331                      [Sports]
7054                [Productivity]
2933                       [Tools]
1920                   [Lifestyle]
8115                      [Action]
6628                   [Lifestyle]
9390                [Productivity]
7816                       [Tools]
3985                   [Education]
4877            [News & Magazines]
10487                      [Tools]
2478                   [Lifestyle]
2546                       [Tools]
3095           [Books & Reference]
4476               [

That's not great either, is it? The text was separated (now each Genre is an array of 1 or 2 values, see for example index 7690, where 'Casual' and 'Education' are separated by a coma), but they are still in the same column, so we are in the same situation. However, split() has an optional argument expand that when it's set to True, instead of creating an array in the same column, will create new columns, and thus returns a dataframe.

In [313]:
stack = df_apps_clean.Genres.str.split(';', expand=True)
stack.sample(30, random_state=30)

0          1
7690                    Casual  Education
5735                     Tools       None
1409                    Sports       None
10306  Video Players & Editors       None
2295                  Shopping       None
5419          Health & Fitness       None
2666                    Action       None
6993                    Social       None
2511                 Education       None
10347            Communication       None
2159              Art & Design       None
7184                     Tools       None
7548             Entertainment       None
5131         Maps & Navigation       None
8331                    Sports       None
7054              Productivity       None
2933                     Tools       None
1920                 Lifestyle       None
8115                    Action       None
6628                 Lifestyle       None
9390              Productivity       None
7816                     Tools       None
3985                 Education       None
4877          News & Magazines       None
10487                    Tools       None
2478                 Lifestyle       None
2546                     Tools       None
3095         Books & Reference       None
4476             Communication       None
4942              Productivity       None

So now we're getting somewhere.. We have 2 columns (that are called "0" and "1" but you can think of them as "Main Genre" and "Additional Genre".. Notice how some values have None there (they already had 1 segment) but the ones that had 2 segments, now have 1 in each column.

Cool, but how do we count them?... One option would be call .value_counts() in the column "0", then do the same for column "1", and add up the values somehow... But there's a better option.

**Step 2 - STACK:** Wouldn't it be easier to count if we could put column "1" below column "0"? In other words, if we could stack them? Surprise, surprise, that's exactly what the stack() function does!

In [314]:
stack = df_apps_clean.Genres.str.split(';', expand=True).stack()
stack.sample(30)

1634  0              Trivia
3558  0               Tools
6010  0        House & Home
2578  0          Simulation
7677  0        Productivity
5615  0          Simulation
1974  0             Finance
4471  0         Educational
8842  0              Sports
3042  0             Medical
8640  0            Business
8096  0           Lifestyle
5831  0             Finance
4486  0     Auto & Vehicles
4499  0       Entertainment
7619  0     Personalization
796   0       Communication
4621  0            Business
1944  0     Personalization
3884  0       Entertainment
9992  0              Action
3016  0             Medical
8839  0              Casino
4496  0    News & Magazines
7345  0           Education
4343  0              Casual
9101  0       Entertainment
4238  0     Personalization
4407  1           Education
845   0     Personalization
dtype: object

Notice how .stack() gave us back a Series from a DataFrame, since it took 2 columns and made it one. **ALSO NOTE** where we see values "0" and "1" that's NOT a column! The 0 and 1 values are INNER-indexes, as mentioned before, we're dealing with a Series now.

In [320]:
stack.index

MultiIndex([(   21, 0),
            (   28, 0),
            (   47, 0),
            (   82, 0),
            (   99, 0),
            (  114, 0),
            (  126, 0),
            (  128, 0),
            (  141, 0),
            (  151, 0),
            ...
            (10807, 0),
            (10813, 0),
            (10817, 0),
            (10818, 0),
            (10822, 0),
            (10824, 0),
            (10828, 0),
            (10829, 0),
            (10831, 0),
            (10835, 0)],
           length=8564)

Since stack it's a Series, we can FINALLY call the .value_counts() on it and get the count by individual Genre.

In [315]:
num_genres = stack.value_counts()
num_genres

Tools                      719
Education                  587
Entertainment              498
Action                     304
Productivity               301
Personalization            298
Lifestyle                  298
Finance                    296
Medical                    292
Sports                     270
Photography                263
Business                   262
Communication              258
Health & Fitness           245
Casual                     216
News & Magazines           204
Social                     203
Simulation                 200
Travel & Local             187
Arcade                     185
Shopping                   180
Books & Reference          171
Video Players & Editors    150
Dating                     134
Puzzle                     124
Maps & Navigation          118
Role Playing               111
Racing                     103
Action & Adventure          96
Strategy                    95
Food & Drink                94
Educational                 93
Adventur

If we compare this with the first count we notice how every row now has only ONE Genre, for instance "Casual" doesn't have the ";Education" part anymore. Also, we can se now that there are 53 genres.

In [319]:
len(num_genres)

53

# Colour Scales in Plotly Charts - Competition in Genres

**Challenge**: Can you create this chart with the Series containing the genre data?

<img src=https://imgur.com/DbcoQli.png width=400>

Try experimenting with the built in colour scales in Plotly. You can find a full list [here](https://plotly.com/python/builtin-colorscales/).

* Find a way to set the colour scale using the color_continuous_scale parameter.
* Find a way to make the color axis disappear by using coloraxis_showscale.

In [317]:
bar = px.bar(x=num_genres.index[:15],
             y=num_genres.values[:15],
             color=num_genres.values[:15],
             color_continuous_scale='Agsunset',
             title='Top Genres',
             hover_name=num_genres.index[:15])
bar.update_layout(xaxis_title='Genre', yaxis_title='Number of Apps', coloraxis_showscale=False)
bar.show()

# Grouped Bar Charts: Free vs. Paid Apps per Category

Now that we’ve looked at the total number of apps per category and the total number of apps per genre, let’s see what the split is between free and paid apps.

In [322]:
df_apps_clean['Type'].value_counts()

Free    7595
Paid     589
Name: Type, dtype: int64

We see that the majority of apps are free on the Google Play Store. But perhaps some categories have more paid apps than others. Let’s investigate. We can group our data first by Category and then by Type. Then we can add up the number of apps per each type. Using as_index=False we push all the data into columns rather than end up with our Categories as the index.

In [330]:
df_free_vs_paid = df_apps_clean.groupby(['Category', 'Type'], as_index=False).agg({'App': pd.Series.count})
df_free_vs_paid

Category  Type  App
0      ART_AND_DESIGN  Free   58
1      ART_AND_DESIGN  Paid    3
2   AUTO_AND_VEHICLES  Free   72
3   AUTO_AND_VEHICLES  Paid    1
4              BEAUTY  Free   42
..                ...   ...  ...
56   TRAVEL_AND_LOCAL  Paid    8
57      VIDEO_PLAYERS  Free  144
58      VIDEO_PLAYERS  Paid    4
59            WEATHER  Free   65
60            WEATHER  Paid    7

[61 rows x 3 columns]

**Challenge**: Use the plotly express bar [chart examples](https://plotly.com/python/bar-charts/#bar-chart-with-sorted-or-ordered-categories) and the [.bar() API reference](https://plotly.com/python-api-reference/generated/plotly.express.bar.html#plotly.express.bar) to create this bar chart:

<img src=https://imgur.com/LE0XCxA.png>

You'll want to use the `df_free_vs_paid` DataFrame that you created above that has the total number of free and paid apps per category.

See if you can figure out how to get the look above by changing the `categoryorder` to 'total descending' as outlined in the documentation here [here](https://plotly.com/python/categorical-axes/#automatically-sorting-categories-by-name-or-total-value).

In [340]:
g_bar = px.bar(data_frame=df_free_vs_paid,
               x='Category',
               y='App',
               color='Type',
               title='Free vs Paid Apps by Category',
               barmode='group')
g_bar.update_layout(xaxis_title='Category',
                    yaxis_title='Number of Apps',
                    xaxis={'categoryorder':'total descending'},
                    yaxis=dict(type='log'))
g_bar.show()

What we see is that while there are very few paid apps on the Google Play Store, some categories have relatively more paid apps than others, including Personalization, Medical and Weather. So, depending on the category you are targeting, it might make sense to release a paid-for app.

But this leads to many more questions:

- How much should you charge? What are other apps charging in that category?
- How much revenue could you make?
- And how many downloads are you potentially giving up because your app is paid?

Let’s try and answer these questions with some Box plots. Box plots show us some handy descriptive statistics in a graph - things like the median value, the maximum value, the minimum value, and some quartiles. Here’s what we’re after:

# Plotly Box Plots: Lost Downloads for Paid Apps

**Challenge**: Create a box plot that shows the number of Installs for free versus paid apps. How does the median number of installations compare? Is the difference large or small?

Use the [Box Plots Guide](https://plotly.com/python/box-plots/) and the [.box API reference](https://plotly.com/python-api-reference/generated/plotly.express.box.html) to create the following chart.

<img src=https://imgur.com/uVsECT3.png>


A box plot is a statistical representation of the distribution of a variable through its quartiles. The ends of the box represent the lower and upper quartiles, while the median (second quartile) is marked by a line inside the box.

In [355]:
box = px.box(df_apps_clean,
             x='Type',
             y='Installs',
             title='How Many Downloads are Paid Apps Giving Up?',
             color='Type',
             notched=True,)
box.update_layout(yaxis=dict(type='log'))
box.show()

From the hover text in the chart, we see that the median number of downloads for free apps is 500,000, while the median number of downloads for paid apps is around 5,000! This is massively lower.

But does this mean we should give up on selling a paid app? Let’s see how much revenue we would estimate per category.

# Plotly Box Plots: Revenue by App Category

**Challenge**: See if you can generate the chart below:

<img src=https://imgur.com/v4CiNqX.png>

Looking at the hover text, how much does the median app earn in the Tools category? If developing an Android app costs $30,000 or thereabouts, does the average photography app recoup its development costs?

Hint: I've used 'min ascending' to sort the categories.

In [365]:
df_paid_apps = df_apps_clean[df_apps_clean['Type'] == 'Paid']
box2 = px.box(df_paid_apps, x='Category',
              y='Revenue_Estimate',
              title='How Much Can Paid Apps Earn?')
box2.update_layout(yaxis_title='Paid App Ballpark Revenue',
                   yaxis=dict(type='log'),
                   xaxis={'categoryorder':'min ascending'})
box2.show()

<img src="https://img-c.udemycdn.com/redactor/raw/2020-10-11_13-55-50-fba0062772e999739ca28119fd0cecda.png">

If an Android app costs \$30,000 to develop, then the average app in very few categories would cover that development cost. The median paid photography app earned about $20,000. Many more app’s revenues were even lower - meaning they would need other sources of revenue like advertising or in-app purchases to make up for their development costs. However, certain app categories seem to contain a large number of outliers that have much higher (estimated) revenue - for example in Medical, Personalisation, Tools, Game, and Family.

So, if you were to list a paid app, how should you price it? To help you decide we can look at how your competitors in the same category price their apps.

# How Much Can You Charge? Examine Paid App Pricing Strategies by Category

**Challenge**: What is the median price price for a paid app? Then compare pricing by category by creating another box plot. But this time examine the prices (instead of the revenue estimates) of the paid apps. I recommend using `{categoryorder':'max descending'}` to sort the categories.

In [366]:
df_paid_apps['Price'].median()

2.99

In [369]:
box3 = px.box(df_paid_apps, x='Category', y='Price', title='Price Per Category')
box3.update_layout(xaxis={'categoryorder':'max descending'}, yaxis=dict(type='log'), yaxis_title='Paid App Price')
box3.show()

<img src="https://img-c.udemycdn.com/redactor/raw/2020-10-11_13-58-35-1c571e911d585782aabbb42b2bb96baf.png">

The median price for an Android app is \$2.99. However, some categories have higher median prices than others. This time we see that Medical apps have the most expensive apps as well as a median price of \$5.49. In contrast, Personalisation apps are quite cheap on average at \$1.49. Other categories which higher median prices are Business (\$4.99) and Dating (\$6.99). It seems like customers who shop in these categories are not so concerned about paying a bit extra for their apps.